### PR TITLE
SE-534-Modify-ERC20-and-ERC721-handlers-to-allow-application-of-rules-based-on-the-user-action

### PR DESCRIPTION
**Pull request by ShaneDuncan602**

### CODE REVIEW

File: docs/userGuides/rules/ACCOUNT-PURCHASE-RULE.md [hunk 1]
```diff
@@ -76,17 +76,20 @@ The collection of these tagged sub-rules plus the startingTime composes an accou
 
 The rule will be evaluated with the following logic:
 
-1. The token handler decides if the transfer is a Purchase (user perspective). Only if it is, it continues with the next steps.
-2. The account is passed to the protocol with all the tags it has registered to its address in the application manager.
-3. The processor receives these tags along with the ID of the account-purchase-controller rule set in the token handler. 
-4. The processor retrieves the sub-rule associated with each tag.
-5. The processor evaluates whether the rule is active based on the `starting timestamp`. If it is not active, the rule aborts the next steps, and returns zero as the accrued cumulative purchases value.
-6. The processor evaluates whether the current time is within a new period.
+1. The handler determines if the rule is active from the supplied action. If not processing does not continue past this step.
```
Comment (TJ-Everett) [inline docs/userGuides/rules/ACCOUNT-PURCHASE-RULE.md]: I think we should have a comma after "If not".

File: docs/userGuides/rules/ACCOUNT-SELL-RULE.md [hunk 2]
```diff
@@ -75,14 +75,17 @@ The collection of these tagged sub-rules plus the startingTime composes an accou
...
 
-1. The token handler decides if the transfer is a Sell (user perspective). Only if it is, it continues with the next steps.
-2. The account being evaluated passes to the protocol all the tags it has registered to its address in the application manager.
-3. The processor receives these tags along with the ID of the account-sell-controller rule set in the token handler. 
-4. The processor then tries to retrieve the sub-rule associated with each tag.
-5. The processor evaluates whether each sub-rule's period is active (if the current time is within `period` from the `starting timestamp`). If it is not within the period, it sets the cumulative sales to the current sale amount. If it is within the period, the processor adds the current sale amount to the accrued sale amount for the rule period.   
-6. The processor then checks if the cumulative sales amount is greater than the `sell amount` defined in the rule. If true, the transaction reverts. 
-7. Steps 4 and 5 are repeated for each of the account's tags.
-8. Return the cumulative sales amount.
+1. The handler determines if the rule is active from the supplied action. If not processing does not continue past this step.
```
Comment (TJ-Everett) [inline docs/userGuides/rules/ACCOUNT-SELL-RULE.md]: Same here.

File: docs/userGuides/rules/ADMIN-WITHDRAWAL-RULE.md [hunk 3]
```diff
@@ -57,9 +57,12 @@ Since this rule is intended to apply specifically to rule bypass accounts, there
 
 The rule will be evaluated with the following logic:
 
-1. The asset handler checks if the transfer of tokens is from an app administrator account. If it is not, the rule evaluation is skipped.
-2. The handler sends the amount of tokens being transferred, the current balance of the app administrator account, and the ruleId to the protocol's rule processor.
-3. The rule processor calculates what the final balance of the administrator account would be if the transaction succeeds. If the final balance calculated is less than the minimum balance specified in the rule, the transaction reverts.
+1. The handler determines if the rule is active from the supplied action. If not processing does not continue past this step.
```
Comment (TJ-Everett) [inline docs/userGuides/rules/ADMIN-WITHDRAWAL-RULE.md]: Same here.

File: docs/userGuides/rules/ALLOW-DENY-ORACLE.md [hunk 4]
```diff
@@ -55,13 +55,16 @@ struct OracleRuleS {
 
 The rule will be evaluated with the following logic (this logic will be evaluated for each oracle rule associated with the token):
 
-1. The processor will receive the ID of the allow-oracle rule set in the application handler. 
-2. The processor will receive the address that is to be checked in the oracle.
-3. The processor will determine the type of oracle based on the rule id. 
-4. The processor will then call the oracle address to check if the address to be checked is on the oracle's list: 
+1. The handler determines if the rule is active from the supplied action. If not processing does not continue past this step.
```
Comment (TJ-Everett) [inline docs/userGuides/rules/ALLOW-DENY-ORACLE.md]: Same here.

File: docs/userGuides/rules/MIN-MAX-ACCOUNT-BALANCE.md [hunk 5]
```diff
@@ -67,12 +67,15 @@ A minumum-maximum-account-balance rule must have at least one sub-rule. There is
 
 The rule will be evaluated with the following logic:
 
-1. The receiver account and the sender account being evaluated pass all the tags they have registered to their addresses in the application manager to the protocol.
-2. The processor receives these tags along with the ID of the minumum-maximum-account-balance rule set in the token handler. 
-3. The processor tries to retrieve the sub-rule associated with each tag.
-4. The processor evaluates if the final balance of the sender account would be less than the`minimum` in the case of the transaction succeeding. If yes, the transaction reverts.
-5. The processor evaluates if the final balance of the receiver account would be greater than the `maximum` in the case of the transaction succeeding. If yes, the transaction reverts.
-6. Step 4 and 5 are repeated for each of the account's tags. 
+1. The handler determines if the rule is active from the supplied action. If not processing does not continue past this step.
```
Comment (TJ-Everett) [inline docs/userGuides/rules/MIN-MAX-ACCOUNT-BALANCE.md]: Same here.

File: docs/userGuides/rules/MINIMUM-ACCOUNT-BALANCE-BY-DATE.md [hunk 6]
```diff
@@ -67,11 +67,14 @@ A minimum-balance-by-date rule must have at least one sub-rule. There is no maxi
 
 The rule will be evaluated with the following logic:
 
-1. The account being evaluated passes all the tags it has registered to their addresses in the application manager to the protocol.
-2. The processor receives these tags along with the ID of the minimum-balance-by-date rule set in the token handler. 
-3. The processor tries to retrieve the sub-rule associated with each tag.
-4. The processor evaluates whether each sub-rule's hold period is still active (if the current time is within `hold period` from the `starting timestamp`). If it is, the processor then evaluates if the final balance of the account would be less than the `hold amount` in the case of the transaction succeeding. If yes, then the transaction reverts.
-5. Step 4 is repeated for each of the account's tags. 
+1. The handler determines if the rule is active from the supplied action. If not processing does not continue past this step.
```
Comment (TJ-Everett) [inline docs/userGuides/rules/MINIMUM-ACCOUNT-BALANCE-BY-DATE.md]: Same here.

File: docs/userGuides/rules/MINIMUM-HOLD-TIME.md [hunk 7]
```diff
@@ -43,9 +43,12 @@ Additionally, each starting unix timestamp for the ownership of the tokenId is s
 
 The rule will be evaluated with the following logic:
 
-1. The handler evaluates the account's `ownershipStart` to check that it is greater than zero.
-2. The handler passes the account's `ownershipStart` and `minimumHoldTimeHours` to the processor. 
-3. The Processor evaluates if the current time minus `ownershipStart` is less than `minimumHoldTimeHours`. If it is the transaction reverts.
+1. The handler determines if the rule is active from the supplied action. If not processing does not continue past this step.
```
Comment (TJ-Everett) [inline docs/userGuides/rules/MINIMUM-HOLD-TIME.md]: Same here

File: docs/userGuides/rules/MINIMUM-TRANSACTION.md [hunk 8]
```diff
@@ -50,10 +50,12 @@ The minimum-transaction rules are stored in a mapping indexed by ruleId(uint32)
 
 The rule will be evaluated with the following logic:
 
-1. The processor receives the ID of the minimum-transaction rule set in the token handler. 
-2. The processor receives the `amount` of tokens from the handler.
-3. The processor evaluates the `amount` against the rule `minTransferAmount` and reverts if the `amount` less than the rule minimum. 
+1. The handler determines if the rule is active from the supplied action. If not processing does not continue past this step.
```
Comment (TJ-Everett) [inline docs/userGuides/rules/MINIMUM-TRANSACTION.md]: and here.

File: docs/userGuides/rules/SUPPLY-VOLATILITY.md [hunk 9]
```diff
@@ -56,12 +56,15 @@ These rules are stored in a mapping indexed by ruleId(uint32) in order of creati
 
 The rule will be evaluated with the following logic:
 
-1. The asset handler keeps track of the total supply per period, the net supply increase/decrease per period, and the date of the last mint/burn transaction.
-2. The asset handler first checks if the transaction is a *minting* or *burning* operation. If it is, then it sends the aforementioned data to the rule processor with the rule Id, and the amount of tokens being minted/burned in the transaction.
-3. The rule processor evaluates if the current transaction is in a new period or part of the same period as the last burning/minting transactions. 
+1. The handler determines if the rule is active from the supplied action. If not processing does not continue past this step.
```
Comment (TJ-Everett) [inline docs/userGuides/rules/SUPPLY-VOLATILITY.md]: and here.

File: docs/userGuides/rules/TOKEN-PERCENTAGE-PURCHASE.md [hunk 10]
```diff
@@ -56,18 +56,21 @@ struct PctPurchaseRuleS {
 
 The rule will be evaluated with the following logic:
 
-1. The token handler decides if the transfer is a Purchase (user perspective). Only if it is, it continues with the next steps.
-2. The processor receives the ID of the token-percentage-purchase rule set in the asset handler. 
-3. The processor receives the current total purchased within period, token A amount (total amount of token A being transferred in the current transaction), previous purchase time, and token's total supply from the handler.
-4. The processor evaluates whether the rule has a set total supply or uses the token's total supply provided by the handler set at the beginning of every new `period`.  
-5. The processor evaluates whether the current time is within a new `period`.
+1. The handler determines if the rule is active from the supplied action. If not processing does not continue past this step.
```
Comment (TJ-Everett) [inline docs/userGuides/rules/TOKEN-PERCENTAGE-PURCHASE.md]: and here

File: docs/userGuides/rules/TOKEN-PERCENTAGE-SELL.md [hunk 11]
```diff
@@ -56,18 +56,21 @@ struct PctSellRuleS {
 
 The rule will be evaluated with the following logic:
 
-1. The token handler decides if the transfer is a Sell (user perspective). Only if it is, it continues with the next steps.
-2. The processor receives the ID of the token-percentage-sell rule set in the asset handler. 
-3. The processor receives the current total sold within period, token A amount (total amount of token A being transferred in the current transaction), previous sell time, and token's total supply from the handler.
-4. The processor evaluates whether the rule has a set total supply or uses the token's total supply provided by the handler set at the beginning of every new `period`.  
-5. The processor evaluates whether the current time is within a new `period`.
+1. The handler determines if the rule is active from the supplied action. If not processing does not continue past this step.
```
Comment (TJ-Everett) [inline docs/userGuides/rules/TOKEN-PERCENTAGE-SELL.md]: same here

File: docs/userGuides/rules/TOKEN-TRANSFER-VOLUME.md [hunk 12]
```diff
@@ -56,18 +56,21 @@ The token-transfer-volume rules are stored in a mapping indexed by ruleId(uint32
 
 The rule will be evaluated with the following logic:
 
-1. The processor receives the ID of the token-transfer-volume rule set in the token handler. 
-2. The processor receives the current `transfer volume`, `last transfer time`, `amount` and token's total supply from the handler.
-3. The processor evaluates whether the rule has a set total supply or use the token's total supply provided by the handler set at the beginning of every new `period`. 
-4. The processor evaluates whether the rule is active based on the `starting timestamp`. If it is not active, the rule evaluation skips the next steps, and simply returns the `transfer volume` value.
-5. The processor evaluates whether the current time is within a new `period`.
+1. The handler determines if the rule is active from the supplied action. If not processing does not continue past this step.
```
Comment (TJ-Everett) [inline docs/userGuides/rules/TOKEN-TRANSFER-VOLUME.md]: same

File: docs/userGuides/rules/TRANSFER-COUNTER.md [hunk 13]
```diff
@@ -63,11 +63,12 @@ A transfer-counter rule must have at least one sub-rule. There is no maximum num
 
 The rule will be evaluated in the following way:
 
-1. The collection being evaluated will pass to the protocol all the tags it has registered to its address in the application manager.
-2. The processor will receive these tags along with the ID of the transfer-counter rule set in the token handler.
-3. The processor will then try to retrieve the sub-rule associated with each tag.
-4. The processor will evaluate whether the trade is within a new period. If no, the rule will continue to the trades per day check (step 5). If yes, the rule will return `tradesInPeriod` of 1 (current trade) for the token Id. 
-5. The processor will evaluate if the total number of trades within the period plus the current trade would be more than the amount of trades allowed per day by the rule in the case of the transaction succeeding. If yes (trades will exceed allowable trades per day), then the transaction will revert.
+1. The handler determines if the rule is active from the supplied action. If not processing does not continue past this step.
```
Comment (TJ-Everett) [inline docs/userGuides/rules/TRANSFER-COUNTER.md]: same here

File: src/client/token/ERC20/ProtocolERC20Handler.sol [hunk 14]
```diff
@@ -116,24 +104,25 @@ contract ProtocolERC20Handler is Ownable, ProtocolHandlerCommon, ProtocolHandler
      * @param _from address of the from account
      * @param _to address of the to account
      * @param _amount number of tokens transferred
+     * @param action if selling or buying (of ActionTypes type)
      */
-    function _checkNonTaggedRules(address _from, address _to, uint256 _amount) internal {
-        if (minTransferRuleActive) ruleProcessor.checkMinTransferPasses(minTransferRuleId, _amount);
+    function _checkNonTaggedRules(address _from, address _to, uint256 _amount, ActionTypes action) internal {
+        if (minTransfer[action].active) ruleProcessor.checkMinTransferPasses(minTransfer[action].ruleId, _amount);
```
Comment (TJ-Everett) [inline src/client/token/ERC20/ProtocolERC20Handler.sol:110]: These action checks are so smart. Such a drastic improvement!

**Review by mpetersoCode55 — APPROVED** [review]

File: src/client/token/ERC20/ProtocolERC20Handler.sol (hunk 14 above)
Comment (oscarsernarosero) [reply src/client/token/ERC20/ProtocolERC20Handler.sol:110]: Agree. This is super nice

File: src/protocol/economic/ruleProcessor/ActionEnum.sol [hunk 15]
```diff
@@ -10,5 +10,7 @@ enum ActionTypes {
     PURCHASE,
     SELL,
     TRADE,
-    INQUIRE
+    MINT,
+    BURN,
+    P2P_TRANSFER
```
Comment (oscarsernarosero) [inline src/protocol/economic/ruleProcessor/ActionEnum.sol:14]: Im still confused about the TRADE action. I think, at this point, we should get rid of it.

Comment (oscarsernarosero) [reply src/protocol/economic/ruleProcessor/ActionEnum.sol:14]: I used recently in the trading rules, but P2P_TRANSFER better describes what I though TRADE was.

File: test/util/TestCommon.sol [hunk 16]
```diff
@@ -419,4 +420,11 @@ abstract contract TestCommon is Test, GenerateSelectors, TestArrays {
     function _createOracleDenied() public returns (OracleDenied _oracleDenied){
         return new OracleDenied(); 
     }
+
+    function _createActionsArray() public pure returns(ActionTypes[] memory){
+        ActionTypes[] memory actionTypes = new ActionTypes[](1);
+        actionTypes[0] = ActionTypes.P2P_TRANSFER;
+        return actionTypes;
+    }
```
Comment (oscarsernarosero) [inline test/util/TestCommon.sol:428]: Love it

**Review by TJ-Everett — APPROVED** [review]
Looks great Shane! Awesome work.

**Review by oscarsernarosero — APPROVED** [review]
Looks amaaaaazing Shane. I just have a super minor suggestion now that I see what you've done.

File: src/protocol/economic/ruleProcessor/ActionEnum.sol (hunk 15 above)
Comment (ShaneDuncan602) [reply src/protocol/economic/ruleProcessor/ActionEnum.sol:14]: agreed

**Pull request merged**